Add mongo sync

speedstyle opened the pull request.

speedstyle committed:
Add mongo sync

speedstyle committed:
Integrate mongo update into bot

speedstyle committed:
dictify add_steam64_ids

theo-brown commented on database/mongo.py:
```diff
@@ -1,28 +1,62 @@
+import aiofiles, aiofiles.os
+from os import getenv
+from discord.ext import tasks, commands as cmds
+from . import users
```
Please could you make imports absolute rather than relative?

theo-brown commented on database/mongo.py:
```diff
@@ -1,28 +1,62 @@
+    @watch.before_loop
+    async def get_token(self):
+        try:
+            async with aiofiles.open('resume', 'rb') as tokenfile:
```
Please could you put the `resume` token in the `database` module?

speedstyle replied:
Done in 68e8a70

theo-brown commented on database/mongo.py:
```diff
@@ -1,28 +1,62 @@
+    async def log_error(self, error):
+        info = await self.bot.application_info()
+        await info.owner.send(
+            f'**{type(error).__name__}**\n{error}'
```
Please could you make this explicit as to where the error came from? It should be clear that it comes from this module, and the exact line/command that raised the error would also be super useful.

speedstyle replied:
If I replace this with `logging.error`, we could later make a `LoggingHandler` that DMs you?

theo-brown replied:
Is there any way that we could make it raise something that's handled by `on_command_error`? Currently `on_command_error` sends the error message to a log channel and then raises `logging.error`.

speedstyle replied:
Since it's not part of a command, no exception can propagate to `on_command_error`. In fact that function wouldn't work here, since it uses `ctx.author`/`ctx.message`/`ctx.channel` which don't make any sense for a long-running internal task.

I think it's best to `logging.error` in both places and use a `LoggingHandler` to send to the log channel? `on_command_error` should probably also tell the user what happened, similar to [what I've done for ratingbot](https://github.com/mxbi/popskill/blob/20ff6675ee1375bcf6b2d857a5aeba269a36012d/discord_app.py#L43).

speedstyle replied:
I've replaced with logging.error for now, I'll work on a logging handler for a future PR

theo-brown replied:
Yep, sounds like this is the best plan

theo-brown commented on database/users.py:
```diff
@@ -34,7 +34,7 @@ async def add_steam64_ids(users):
         "INSERT OR REPLACE INTO users(user_id, steam64_id)"
         " VALUES (?, ?)",
-        ((user['discord_id'], user['steam_id']) for user in users)
+        ((discord, steam) for discord,steam in users.items())
```
Please could you add a type hint for users, given it now should be a dict?

speedstyle replied:
Done in 68e8a70

theo-brown commented on database/__init__.py:
```diff
@@ -20,3 +21,4 @@ async def async_init(self):
 
 def setup(bot):
     bot.db = DatabaseExtension(bot)
+    bot.add_cog(MongoCog(bot))
```
This seems to be a different way of adding cogs than what we've previously done, was that intentional?

speedstyle replied:
I'm taking the `Cog` subclass and calling `bot.add_cog(MyCog(bot))`, as before. The `setup` function isn't in the same file as the cog, since the mongo cog is part of the database extension. I guess I could add a `setup` to `mongo` and add `load_extension(database.mongo)`? but this way made most sense to me

theo-brown replied:
Very sorry, that's just me completely misremembering what we did elsewhere.

theo-brown replied:
That's entirely my mistake, I forgot how we did this in the rest of the code.

theo-brown commented on database/__init__.py:
```diff
@@ -1,4 +1,5 @@
 import aiosqlite
+from .mongo import MongoCog
```
Please could you make this import absolute?

theo-brown reviewed (changes requested):
Good stuff! Could you make a couple of changes?

speedstyle committed:
Fixes for #16 review

Parameterized/moved resume file, typed some database.users code and fixed a bug in !update.
